pkg/cloud/openstack/clients/machineservice.go

```diff
@@ -275,6 +277,83 @@ func CreatePort(is *InstanceService, name string, net ServerNetwork, securityGro
 	return *newPort, nil
 }
 
+func isDuplicate(list []string, name string) bool {
+	if list == nil || len(list) == 0 {
+		return false
+	}
+	for _, element := range list {
+		if element == name {
+			return true
+		}
+	}
+	return false
+}
+
+func GetSecurityGroups(is *InstanceService, sg_param []openstackconfigv1.SecurityGroupParam) ([]string, error) {
```

Comment:
You may want to take a look at what I did for the `subnets` filter in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/2ee134e451af7c42c81ddd72693a320bff9b285c